### PR TITLE
SQL Relationships - UI and backend

### PR DESCRIPTION
**Pull request by mike12345567**
## Description
This is the combination of two branches, `feature/opinionated-relationship-ui` and `feature/sql-relationships`. I'm merging them both as the UI work depends on the backend work to be useful, so it makes sense to bring them in together.

The concept for this piece of work was to create the same functionality that we have with internal relationships, but make it usable for SQL relationships. This doesn't offer the full power of wide joining capabilities that you usually get in SQL but it offers a fairly consistent and opinionated way of looking up that doesn't require any queries.

The idea with this work is that basic viewing of relationships in tables and CRUD for creating relationships, can be easily executed without writing specific queries for each - this is specifically very good at many to many relationships in SQL were this would involve a somewhat complex process for managing the junction table.

A key point of this is that the API is nearly identical to our internal relationship API, this comes with some negatives in that our internal relationships are not as powerful as the full capabilities of SQL relationships, but it means that everything that you can do with internal tables can be completely replicated (and efficiently) with external SQL databases. If we did it in such a way that the relationships had the full capabilities of SQL relationships (wide rows with multi-layer joining etc) then we would have needed to essentially re-build how data providers interact with the system as all requests could have dynamic schemas - it would have been a much, much bigger project.

In the backend most of the work went into the `external` and `ExternalRequest.ts` components of the row API controllers, it is mostly an extension of the existing system, with a new `relationships` component of the JSON structure that we use for querying from external datasources. Most of this has been defined in Typescript so that for the future the structure is always known.

The actual functionality under the hood is quite performance efficient, only fields required are requested (`SELECT *` is never used) and joins are using sparingly in the form of `JOIN LEFT` as we always put the from table on the left hand side of statements.

In terms of improvements that could be made there are the following:
1. Bulk deletions, right now these are done one by one, however it would be easy enough to change this over to use a `oneOf` filter instead.
2. Transactions, right now a lot of the many-to-many operations are not atomic as the system doesn't really have transaction capabilities right now, this could be added in the future however, adding a function to the SQL library to start a transaction, and then end. The concept of the `ExternalRequest` structure helps with this as all requests will be made as part of the existence of the external request object, so essentially as part of a constructor and faked destructor we could start/stop a transaction, meaning the requests wouldn't actually be made until they had all been gathered at the end.
3. Right now relationships are not composite key aware, if a foreign key is formed from a composite key the system cannot handle this.

To really demonstrate the feature I felt it was better to simply show a quick video of how quickly you can go from no connection, to a fully connected up relationship experience. I didn't build any sort of app as the API is the same, everything that appears in the backend shows its identical, so everything you can do with internal tables can be done in actual apps with SQL tables.

_Loom below_
https://www.loom.com/share/63862c2a15b142b1b918407c7775123a

### UI
The new buttons added to the external data source plus screens:
image

Create relationship modal:
image

Display columns modal:
image

Table experience:
image

### CODE REVIEW

**Comment by mike12345567** [comment]
recheck

**Comment by joebudi** [comment]
Great work. I know how difficult this was to implement so credit to Michael and Marty.

File: packages/builder/src/pages/builder/app/[application]/data/datasource/[selectedDatasource]/CreateEditRelationship/CreateEditRelationship.svelte [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+      <Select
+        label={`Foreign Key (${toTable?.name})`}
+        options={Object.keys(toTable?.schema)}
+        bind:value={fromRelationship.fieldName}
+      />
+    {/if}
+  </div>
+
+  <div slot="footer">
+    {#if originalFromName !== null}
```
Comment (aptkingston) [inline packages/builder/src/pages/builder/app/[application]/data/datasource/[selectedDatasource]/CreateEditRelationship/CreateEditRelationship.svelte:180]: Just need to swap this to `!=` to catch `undefined` - I noticed the delete button was still visible when creating a relationship and the extra equals here is the culprit.

File: packages/builder/src/pages/builder/app/[application]/data/datasource/[selectedDatasource]/CreateEditRelationship/CreateEditRelationship.svelte [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+      <Button warning text on:click={deleteRelationship}>Delete</Button>
+    {/if}
+  </div>
+</ModalContent>
+
+<style>
+  .table-selector {
+    display: grid;
+    grid-template-columns: repeat(5, 1fr);
```
Comment (aptkingston) [inline packages/builder/src/pages/builder/app/[application]/data/datasource/[selectedDatasource]/CreateEditRelationship/CreateEditRelationship.svelte]: There can only be 4 columns here can't there? Might look a bit better to reduce it down to 4 grid columns.

File: packages/builder/src/pages/builder/app/[application]/data/datasource/[selectedDatasource]/index.svelte [hunk 3]
```diff
@@ -92,28 +171,62 @@
...
-          >
+          <div class="table-buttons">
+            {#if plusTables && plusTables.length !== 0}
+              <Button primary on:click={openDisplayColumnModal}>
+                Update display columns
+              </Button>
+            {/if}
+            <div>
+              <Button primary on:click={updateDatasourceSchema}>
+                Fetch Tables From Database
```
Comment (aptkingston) [inline packages/builder/src/pages/builder/app/[application]/data/datasource/[selectedDatasource]/index.svelte:182]: Super insignificant (and there are still plenty of place in the builder that need updated) but Adobe's recommendation is always to use normal sentence case for buttons - so just capitalise the first letter of the first word and lower case the rest. So just "Fetch tables from database" here, like it is with the display columns button.

File: packages/builder/src/pages/builder/app/[application]/data/datasource/[selectedDatasource]/index.svelte [hunk 4]
```diff
@@ -92,28 +171,62 @@
...
+              <p>→</p>
+            </div>
+          {/each}
+        </div>
+
+        <Divider />
+        <div class="query-header">
+          <Heading size="S">Relationships</Heading>
+          <Button primary on:click={() => openRelationshipModal()}
+            >Create Relationship</Button
```
Comment (aptkingston) [inline packages/builder/src/pages/builder/app/[application]/data/datasource/[selectedDatasource]/index.svelte:206]: Same as other buttons, theme and case just.

**Review by aptkingston — APPROVED** [review]
💯 Massive piece of work, amazing job guys. Next stop: relationships between different databases! 🤯
Just a couple of minor UI things I noticed, but nothing major.

File: packages/builder/src/pages/builder/app/[application]/data/datasource/[selectedDatasource]/CreateEditRelationship/CreateEditRelationship.svelte (hunk 2 above)
Comment (mike12345567) [reply packages/builder/src/pages/builder/app/[application]/data/datasource/[selectedDatasource]/CreateEditRelationship/CreateEditRelationship.svelte]: Removed the styles in general, screenshot of new UI above.

**Review by shogunpurple — APPROVED** [review]
LGTM - great job!

**Pull request merged**